### PR TITLE
F4ever review PR

### PR DESCRIPTION
**Pull request by F4ever**

### CODE REVIEW

File: src/common/registry-csm/fetch/key.constants.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export const REGISTRY_KEY_BATCH_SIZE = 200;
```
Comment (F4ever) [inline src/common/registry-csm/fetch/key.constants.ts:1]: Not used

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -36,7 +36,7 @@ jobs:
       - name: Run tests
         run: yarn test
       - name: Run E2E tests
-        run: yarn test:e2e:docker
+        run: yarn test:e2e
```
Comment (F4ever) [inline .github/workflows/test.yml:39]: - Outdated rest-api.md
    - Invalid interface for ELBlockSnapshot
    - lastChangedBlockHash - not represented in API
    - v1/modules - Module inrterface not full
    - /v1/modules/{module_id}/validators/validator-exits-to-prepare/ if not exist NO should be 404 exception

File: src/staking-router-modules/staking-module-impl-config.ts [hunk 3]
```diff
@@ -1,8 +1,11 @@
 import { CuratedModuleService } from './curated-module.service';
+import { CommunityModuleService } from './community-module.service';
 import { STAKING_MODULE_TYPE } from './constants';
 
-export type StakingModuleImpl = typeof CuratedModuleService;
+export type StakingModuleImpl = typeof CuratedModuleService | typeof CommunityModuleService;
 
 export const config: Record<STAKING_MODULE_TYPE, StakingModuleImpl> = {
```
[the remark this answers is not in the document]
Comment (Amuhar) [reply src/staking-router-modules/staking-module-impl-config.ts:7]: `StakingModuleInterface` defines the type for instances of the classes, not the class references themselves. In config, we need to store references to the class constructors, and typeof ClassName is necessary for this purpose. Interfaces can't represent class constructors.

Using typeof allows us to retrieve instances with this.moduleRef.get<StakingModuleInterface>(impl);
ModuleRef is used for dependency injection in NestJS. It allows us to retrieve an instance of a class at runtime.

File: .github/workflows/test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Amuhar) [reply .github/workflows/test.yml:39]: We can get rid of  rest-api.md 
And just mention how to generate doc in readme

File: README.md [hunk 4]
```diff
@@ -47,9 +47,28 @@ For configuring Grafana, navigate to http://localhost:8000/dashboards. Here, you
 
 To view the list of alerts in Prometheus, visit http://localhost:9090/alerts. For checking the list of fired alerts in Alertmanager, go to http://localhost:9093/#/alerts.
 
+## Update CSM module ABI
+
+Execute the command in the CSM module repository
```
Comment (Amuhar) [inline README.md:52]: make we need to add repo link, it is not clear what do we mean here by CSM module repository

File: src/app/simple-dvt-deploy.e2e-chain.ts [hunk 5]
```diff
@@ -100,7 +100,7 @@ describe('Simple DVT deploy', () => {
     stakingRouterService = moduleRef.get(StakingRouterService);
 
     jest.spyOn(configService, 'get').mockImplementation((path: any) => {
-      if (path === 'LIDO_LOCATOR_ADDRESS') {
+      if (path === 'LIDO_LOCATOR_DEVNET_ADDRESS') {
```
Comment (Amuhar) [inline src/app/simple-dvt-deploy.e2e-chain.ts:103]: how does it work for holesky? 

will it always use in test devnet locator address?

File: src/common/registry-csm/fetch/key-batch.constants.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+/**
+ * keys 96 = 48byte
```
[the remark this answers is not in the document]
Comment (Amuhar) [reply src/common/registry-csm/fetch/key-batch.constants.ts:2]: pair of hexadecimal digits represents one byte of data

File: src/http/common/entities/sr-module.ts [hunk 7]
```diff
@@ -21,69 +21,81 @@ export class StakingModuleResponse implements Omit<SrModuleEntity, 'id' | 'modul
...
     description: 'block.timestamp of the last deposit of the module',
   })
   lastDepositAt: number;
 
   @ApiProperty({
+    required: true,
     description: 'block.number of the last deposit of the module',
-    nullable: true,
   })
   lastDepositBlock: number;
```
Comment (Amuhar) [inline src/http/common/entities/sr-module.ts:89]: TODO: check what contract return

File: src/common/registry-csm/main/validator-registry/validator-registry.service.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { RegistryOperator } from '../../../registry/storage/operator.entity';
+import { AbstractRegistryService } from '../abstract-registry';
+
+export class ValidatorRegistryService extends AbstractRegistryService {
```
Comment (Amuhar) [inline src/common/registry-csm/main/validator-registry/validator-registry.service.ts:4]: we dont need this service

File: src/common/registry-csm/main/constants.ts [hunk 9]
```diff
@@ -0,0 +1 @@
+export const REGISTRY_GLOBAL_OPTIONS_TOKEN = Symbol('registryGlobalOptions');
```
Comment (Amuhar) [inline src/common/registry-csm/main/constants.ts:1]: it looks unsafe use the same value as in  registry

File: src/common/registry-csm/fetch/utils/batches.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
+export const makeBatches = (batchSize: number, offset: number, limit: number) => {
+  if (!Number.isInteger(offset) || offset < 0) {
+    throw new RangeError('offset should be positive integer');
+  }
+  if (!Number.isInteger(limit) || limit < 0) {
```
[the remark this answers is not in the document]
Comment (Amuhar) [reply src/common/registry-csm/fetch/utils/batches.ts:5]: that is for safety

File: src/common/registry-csm/fetch/key-batch.fetch.ts [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    return keys.map((key, chunkIndex) => {
+      const index = startIndex + chunkIndex;
+      return {
+        operatorIndex,
+        index,
+        key: key,
+        depositSignature: signatures[chunkIndex],
+        // TODO: write test
```
Comment (Amuhar) [inline src/common/registry-csm/fetch/key-batch.fetch.ts:54]: did we write test or create task ?

**Comment by Amuhar** [comment]
Some aspects of the code need attention

File: src/common/registry-csm/fetch/utils/batches.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (F4ever) [reply src/common/registry-csm/fetch/utils/batches.ts:5]: for better safety, better `limit < 1`. Because in case `limit == 0` function returns empty list, even there should be some requests

File: src/common/registry-csm/fetch/utils/batches.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
+export const makeBatches = (batchSize: number, offset: number, limit: number) => {
```
Comment (F4ever) [inline src/common/registry-csm/fetch/utils/batches.ts:1]: DRY

File: scripts/prepare-csm-abi.js [hunk 13]
```diff
@@ -0,0 +1,37 @@
+const fs = require('fs');
```
[the remark this answers is not in the document]
Comment (eddort) [reply scripts/prepare-csm-abi.js:1]: It's a script, not a service, we don't test it

File: src/staking-router-modules/contracts/abi/csm.json [hunk 14]
```diff
@@ -0,0 +1,587 @@
+[
```
[the remark this answers is not in the document]
Comment (eddort) [reply src/staking-router-modules/contracts/abi/csm.json:1]: We usually store in a formatted style, but there was no clear agreement.

File: src/common/registry-csm/fetch/operator.fetch.ts [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    if (nodeOperatorRewardAddressChangedEvents.length) {
+      return true;
+    }
+
+    return false;
+  }
+
+  /** return blockTag for finalized block, it need for testing purposes */
+  public getFinalizedBlockTag() {
+    return 'finalized';
```
[the remark this answers is not in the document]
Comment (eddort) [reply src/common/registry-csm/fetch/operator.fetch.ts:60]: This is needed for tests because ganache doesn't have a `finalised` tag

File: src/app/simple-dvt-deploy.e2e-chain.ts (hunk 5 above)
Comment (eddort) [reply src/app/simple-dvt-deploy.e2e-chain.ts:103]: In another review we discussed that it is better to call this variable LIDO_LOCATOR_DEVNET_ADDRESS since it is only used in Devnet

File: src/common/registry-csm/main/key-registry/key-registry.module.ts [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    };
+  }
+
+  static forFeature(options?: RegistryModuleSyncOptions): DynamicModule {
+    return {
+      module: CSMKeyRegistryModule,
+      imports: [...(options?.imports || []), RegistryFetchModule.forFeature(options)],
+      providers: [
+        {
+          provide: REGISTRY_GLOBAL_OPTIONS_TOKEN,
```
[the remark this answers is not in the document]
Comment (eddort) [reply src/common/registry-csm/main/key-registry/key-registry.module.ts:34]: You are right that this can be a problem, but in our case we are using Symbol and Symbol has the property that Symbol(‘registryGlobalOptions’) !== Symbol(‘registryGlobalOptions’)

File: src/common/registry-csm/main/constants.ts (hunk 9 above)
Comment (eddort) [reply src/common/registry-csm/main/constants.ts:1]: You are right that this can be a problem, but in our case we are using Symbol and Symbol has the property that Symbol(‘registryGlobalOptions’) !== Symbol(‘registryGlobalOptions’)

File: src/scripts/constants.ts [hunk 17]
```diff
@@ -0,0 +1,12 @@
+export const CSM_CONTRACT_ADDRESS = '0x26aBc20a47f7e8991F1d26Bf0fC2bE8f24E9eF2A';
```
[the remark this answers is not in the document]
Comment (eddort) [reply src/scripts/constants.ts]: These scripts have been removed, they were only needed for the development of the module

File: src/common/registry-csm/fetch/operator.fetch.ts [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+    return bigNumber.toNumber();
+  }
+
+  /**
+   * fetches finalized operator
+   * @param moduleAddress address of sr module
+   * @param operatorIndex index of sr module operator
+   * @returns used signing keys count, if error happened returns 0 (because of range error)
+   */
+  public async getFinalizedNodeOperatorUsedSigningKeys(moduleAddress: string, operatorIndex: number): Promise<number> {
```
[the remark this answers is not in the document]
Comment (eddort) [reply src/common/registry-csm/fetch/operator.fetch.ts:75]: Your suggestion will break the rest of the algorithm. We use 0, because it will force us to reload all keys in all cases we are interested in.

File: src/common/registry-csm/fetch/operator.fetch.ts [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+
+  /** fetches one operator */
+  public async fetchOne(
+    moduleAddress: string,
+    operatorIndex: number,
+    overrides: CallOverrides = {},
+  ): Promise<RegistryOperator> {
+    const contract = this.getContract(moduleAddress);
+
+    const operator = await contract.getNodeOperator(operatorIndex, overrides as any);
```
[the remark this answers is not in the document]
Comment (eddort) [reply src/common/registry-csm/fetch/operator.fetch.ts:100]: Thanks! we will consider this improvement in the next refactoring.

File: src/common/registry-csm/fetch/key-batch.fetch.ts [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+      throw new Error('fromIndex is greater than or equal to toIndex in CSM module');
+    }
+
+    if (toIndex == null || toIndex === -1) {
+      const operator = await this.operatorsService.fetchOne(moduleAddress, operatorIndex, overrides);
+
+      toIndex = operator.totalSigningKeys;
+    }
+
+    if (usedKeysCount == null || usedKeysCount === -1) {
```
[the remark this answers is not in the document]
Comment (eddort) [reply src/common/registry-csm/fetch/key-batch.fetch.ts:80]: Thanks! we will consider this improvement in the next refactoring.

File: src/common/registry-csm/fetch/key-batch.fetch.ts [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+    fromIndex = 0,
+    toIndex = -1,
+    usedKeysCount = -1,
+    overrides: CallOverrides = {},
+  ): Promise<RegistryKey[]> {
+    if (fromIndex > toIndex && toIndex !== -1) {
+      throw new Error('fromIndex is greater than or equal to toIndex in CSM module');
+    }
+
+    if (toIndex == null || toIndex === -1) {
```
[the remark this answers is not in the document]
Comment (eddort) [reply src/common/registry-csm/fetch/key-batch.fetch.ts:74]: Thanks! we will consider this improvement in the next refactoring.

File: src/common/registry-csm/fetch/utils/batches.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (eddort) [reply src/common/registry-csm/fetch/utils/batches.ts:5]: This function correctly covers the edge cases. This case is shown in the test, there you can also see the expected behaviour.

https://github.com/lidofinance/lido-keys-api/blob/d8c7488b9f934f8597233d5d784223b9b71f8bb4/src/common/registry/fetch/utils/batches.spec.ts#L60

**Comment by eddort** [comment]
Thanks for the review, we will take your comments into consideration when making a plan for future refactoring

**Comment by eddort** [comment]
Thanks for the review, we will take your comments into consideration when making a plan for future refactoring

File: src/common/registry-csm/main/validator-registry/validator-registry.module.ts [hunk 22]
```diff
@@ -0,0 +1,56 @@
+import { DynamicModule, Module } from '@nestjs/common';
+import { RegistryModuleSyncOptions, RegistryModuleAsyncOptions } from '../interfaces/module.interface';
+import { ValidatorRegistryService } from './validator-registry.service';
+import { RegistryStorageModule } from '../../../registry/storage/registry-storage.module';
+import { RegistryFetchModule } from '../../fetch/registry-fetch.module';
+import { REGISTRY_GLOBAL_OPTIONS_TOKEN } from '../constants';
+
+@Module({
```
[the remark this answers is not in the document]
Comment (eddort) [reply src/common/registry-csm/main/validator-registry/validator-registry.module.ts:8]: It has to be done this way because of the framework
